common/stacks/src/main/scala/weco/api/stacks/services/ItemLookup.scala

```diff
@@ -0,0 +1,106 @@
+package weco.api.stacks.services
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.{ElasticError, Index}
+import uk.ac.wellcome.models.Implicits._
+import weco.api.search.elasticsearch.ElasticsearchService
+import weco.catalogue.internal_model.identifiers.{
+  CanonicalId,
+  IdState,
+  SourceIdentifier
+}
+import weco.catalogue.internal_model.work.{Item, Work}
+import weco.catalogue.internal_model.work.WorkState.Indexed
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class ItemLookup(elasticsearchService: ElasticsearchService)(
+  implicit ec: ExecutionContext
+) {
+
+  /** Returns an Item that corresponds to this canonical ID.
+    *
+    * Note: the output from this function may not be consistent, even against
+    * a fixed API index.  Because of merging logic, some item IDs have varying
+    * representations on different works.
+    *
+    */
+  def byCanonicalId(itemId: CanonicalId)(index: Index)
+    : Future[Either[ElasticError, Option[Item[IdState.Identified]]]] = {
+    val searchRequest =
+      search(index)
+        .query(
+          boolQuery.filter(
+            termQuery("data.items.id.canonicalId", itemId.underlying),
+            termQuery(field = "type", value = "Visible")
+          )
+        )
+        .size(1)
+
+    elasticsearchService.findBySearch[Work[Indexed]](searchRequest).map {
+      case Left(err) => Left(err)
+      case Right(works) =>
+        Right(
+          works
+            .flatMap { _.data.items }
+            .collectFirst {
+              // The .asInstanceOf here will be a no-op at runtime, and is just so
+              // the compiler knows this will always be an Item[IdState.Identified]
```

Comment:
I'm surprised the pattern match doesn't tell it this! It definitely does _sometimes_

Reply:
Yeah, it’s one of those things I expected to work. Also, I thought I'd written a much longer comment here, hmm. Will add back in the next PR.